src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java

```diff
@@ -0,0 +1,424 @@
+import org.apache.spark.sql.RowFactory;
+
+import org.apache.spark.sql.types.*;
+import org.apache.spark.sql.Row;
+
+/**
+ * Created by ganeshkumar on 9/29/15.
+ */
+public class SchemaPreservingSolrRDD extends SolrRDD {
+  public static Logger log = Logger.getLogger(SolrRDD.class);
+
+  public SchemaPreservingSolrRDD(String collection) {
+    super("localhost:9983", collection); // assume local embedded ZK if not supplied
+  }
+
+  public SchemaPreservingSolrRDD(String zkHost, String collection) {
+    super(zkHost, collection);
+  }
+
+  public StructType getQuerySchema(SolrQuery query) throws Exception {
+    query.addFilterQuery("__lwcategory_s:schema AND __lwroot_s:root");
+    JavaRDD<SolrDocument> rdd1 = queryShards(sc, query);
+    return readSchema(rdd1.collect().get(0), getSolrClient(zkHost) , collection);
+  }
+
+  public static StructType readSchema(SolrDocument doc, SolrClient Solr, String collection) throws IOException, SolrServerException {
+    List<StructField> fields = new ArrayList<StructField>();
+    StructType st= (StructType) recurseReadSchema(doc, Solr, fields, collection).dataType();
+    return st;
+  }
+
+  public static StructField recurseReadSchema(SolrDocument doc, SolrClient solr, List<StructField> fldr, String collection) throws IOException, SolrServerException {
+    Boolean recurse = true;
+    String finalName = null;
+    for (java.util.Map.Entry<String, Object> field : doc.entrySet()) {
+      String name = field.getKey();
+      Object value = field.getValue();
+      if (name.startsWith("links")) {
+        String id = doc.get(name).toString();
+        if (id != null) {
+          SolrQuery q1 = new SolrQuery("id:" + id);
+          QueryResponse rsp1 = null;
+          try {
+            rsp1 = solr.query(collection,q1);
+          } catch (Exception E) {
+            log.error(E.toString());
+            recurse = false;
+          }
+          if (recurse) {
+            SolrDocumentList docs1 = rsp1.getResults();
+            List<StructField> fld1 = new ArrayList<StructField>();
+            fldr.add(recurseReadSchema(docs1.get(0), solr, fld1, collection));
+          }
+        }
+      }
+      if (name.substring(name.length()-2,name.length()).equals("_s")  && !name.equals("__lwroot_s") && !name.startsWith("links") && !name.equals("__lwcategory_s")) {
+        if (name.substring(0, name.length() - 2).equals("__lwchilddocname")) {
+          finalName = field.getValue().toString();
+        } else {
+            fldr.add(new StructField(name.substring(0, name.length() - 2), getsqlDataType(field.getValue().toString()), true, Metadata.empty()));
+          }
+      }
+
+    }
+    StructField[] farr = new StructField[fldr.size()];
+    farr = fldr.toArray(farr);
+    StructType st2 = new StructType(farr);
+    if (finalName == null) {
+      finalName = "root";
+    }
+    return new StructField(finalName, st2, true,  Metadata.empty());
+  }
+
+  public static DataType getsqlDataType(String s) {
```

Comment:
This seems like a useful function for SolrRDD too .. consider moving this method back up into SolrRDD